src/coreclr/nativeaot/Runtime/unix/PalRedhawkUnix.cpp

```diff
@@ -1212,21 +1212,45 @@ extern "C" void GetSystemTimeAsFileTime(FILETIME *lpSystemTimeAsFileTime)
 
 extern "C" UInt32_BOOL QueryPerformanceCounter(LARGE_INTEGER *lpPerformanceCount)
 {
-    // TODO: More efficient, platform-specific implementation
-    struct timeval tv;
-    if (gettimeofday(&tv, NULL) == -1)
+#if HAVE_CLOCK_GETTIME_NSEC_NP
```

Comment:
Are all these `HAVE_...` constants properly defined here?

It may be best to piggy back on the GCOS interface for this. Look for `GCToOSInterface::` in this file.

Reply:
It looks like the src\coreclr\nativeaot\Runtime\unix\configure.cmake already detects all of these

Reply:
I think it would not build if `HAVE_CLOCK_GETTIME_NSEC_NP` and `CLOCK_MONOTONIC` were not defined on Linux and OSX accordingly.  (I used this change on both)

Reply:
Are you suggesting calling `GCToOSInterface::` from PAL ?   Should it normally be the other way around?

Reply:
The Redhawk PAL is put together in odd ways. We call GCToOSInterface:: from it in several places.

I think that the long-term solution is to move these sort of methods to https://github.com/dotnet/runtime/tree/main/src/native/minipal and avoid the duplication.

Reply:
Alos it looks like `GCToOSInterface::QueryPerformanceCounter` just uses `gettimeofday`. We need something more lightweight here.

https://github.com/dotnet/runtime/blob/cb6fb60dfc253bc13655fafa15686cb864fd0ea2/src/coreclr/gc/unix/gcenv.unix.cpp#L1417

Reply:
The way GC uses the clock should be ok if it is a monotonic counter. I will change that and use from here.

Reply:
I assume it will be a direct call, no roundabout virtual shims in case GC is dynamically loaded.

Reply:
Right, it is not in GC, it is in the env, which is static.

Reply:
When GC is dynamically loaded, it is in the GC. The GC needs to call directly to the GCToOSInterface as many methods are perf sensitive.

Reply:
But there is still a copy in the env that you can use.

Reply:
Looks like GC does not have `HAVE_CLOCK_MONOTONIC`

Reply:
I have added `HAVE_CLOCK_MONOTONIC` detection